zipline/data/data_portal.py

```diff
@@ -0,0 +1,1312 @@
+
+        trading_days: pd.DatetimeIndex
+            The trading range.
+
+        Returns
+        -------
+        list: A list of objects with all relevant attributes populated.
+        All timestamp fields are converted to pd.Timestamps.
+        """
+
+        if self._adjustment_reader is None:
+            return []
+
+        if len(trading_days) == 0:
+            return []
+
+        start_dt = trading_days[0].value / 1e9
+        end_dt = trading_days[-1].value / 1e9
+
+        dividends = self._adjustment_reader.conn.execute(
+            "SELECT * FROM stock_dividend_payouts WHERE sid = ? AND "
+            "ex_date > ? AND pay_date < ?", (int(sid), start_dt, end_dt,)).\
+            fetchall()
+
+        dividend_info = []
+        for dividend_tuple in dividends:
+            dividend_info.append({
+                "declared_date": dividend_tuple[1],
+                "ex_date": pd.Timestamp(dividend_tuple[2], unit="s"),
+                "pay_date": pd.Timestamp(dividend_tuple[3], unit="s"),
+                "payment_sid": dividend_tuple[4],
+                "ratio": dividend_tuple[5],
+                "record_date": pd.Timestamp(dividend_tuple[6], unit="s"),
+                "sid": dividend_tuple[7]
+            })
+
+        return dividend_info
+
+    def contains(self, asset, field):
+        return field in BASE_FIELDS or \
+            (field in self._augmented_sources_map and
+             asset in self._augmented_sources_map[field])
+
+    def get_fetcher_assets(self):
+        """
+        Returns a list of assets for the current date, as defined by the
+        fetcher data.
+
+        Notes
+        -----
+        Data is forward-filled.  If there is no fetcher data defined for day
+        N, we use day N-1's data (if available, otherwise we keep going back).
+
+        Returns
+        -------
+        list: a list of Asset objects.
+        """
+        # return a list of assets for the current date, as defined by the
+        # fetcher source
+        if self._extra_source_df is None:
+            return []
+
+        if self.current_day in self._extra_source_df.index:
+            date_to_use = self.current_day
+        else:
+            # current day isn't in the fetcher df, go back the last
+            # available day
+            idx = self._extra_source_df.index.searchsorted(self.current_day)
+            if idx == 0:
+                return []
+
+            date_to_use = self._extra_source_df.index[idx - 1]
+
+        asset_list = self._extra_source_df.loc[date_to_use]["sid"]
+
+        # make sure they're actually assets
+        asset_list = [asset for asset in asset_list
+                      if isinstance(asset, Asset)]
+
+        return asset_list
+
+
+class DataPortalSidView(object):
+    def __init__(self, asset, portal):
+        self.asset = asset
+        self.portal = portal
+
+    def __getattr__(self, column):
+        return self.portal.get_spot_value(self.asset, column)
```

Comment:
cc: @ssanderson this is where `get_spot_value` is called without a parameter.
The context where this is called is when an algorithm indexes into `data`, e.g. `data[sid(24)].price`.